sycl/unittests/helpers/PiMock.hpp

```diff
@@ -0,0 +1,192 @@
+/// The underlying SYCL platform must be a non-host plaftorm to facilitate
+/// plugin usage.
+///
+/// Simple usage examples would look like this:
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// unittest::PiMock Mock;
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// platform &MockP = Mock.getPlatform();
+/// /*...*/
+/// ```
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// queue Q;
+/// unittest::PiMock Mock(Q);
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// Q.submit(/* expecting mock behavior */);
+/// /*...*/
+/// ```
+class PiMock {
+public:
+  /// Default-constructs PiMock.
+  ///
+  /// Delegates to the device_selector-based constructor by passing
+  /// a default_constructor and assuming a non-host device will be
+  /// available.
+  PiMock() : PiMock(cl::sycl::default_selector{}) {}
+
+  /// Constructs PiMock from a queue.
+  ///
+  /// \param Queue is a reference to a SYCL queue to which
+  ///        the mock redefinitions will apply.
+  explicit PiMock(cl::sycl::queue &Queue)
+      : PiMock(Queue.get_device().get_platform()) {}
+
+  /// Constructs PiMock from a reference to a SYCL platform instance.
+  ///
+  /// A new plugin will be stored into the platform instance, which
+  /// will no longer share the plugin with other platform instances
+  /// within the given context. A separate platform instance will be
+  /// held by the PiMock instance.
+  ///
+  /// \param OriginalPlatform is a reference to a SYCL platform.
+  explicit PiMock(const cl::sycl::platform &OriginalPlatform) {
+    assert(!OriginalPlatform.is_host() && "PI mock isn't supported for host");
+    // Extract impl and plugin handles
+    std::shared_ptr<detail::platform_impl> ImplPtr =
+        detail::getSyclObjImpl(OriginalPlatform);
+    const RT::PiPlugin &OriginalPiPlugin = ImplPtr->getPlugin().getPiPlugin();
+    // Copy the PiPlugin, thus untying our to-be mock platform from other
+    // platforms within the context. Reset our platform to use the new plugin.
+    auto NewPluginPtr = std::make_shared<detail::plugin>(OriginalPiPlugin);
+    ImplPtr->setPlugin(NewPluginPtr);
+    // Extract the new PiPlugin instance by a non-const pointer,
+    // explicitly allowing modification
+    MPiPluginMockPtr = &NewPluginPtr->getPiPlugin();
+    // Save a copy of the platform resource
+    MPlatform = OriginalPlatform;
+  }
+
+  /// Constructs PiMock from a device_selector, provided that
+  /// a non-host device can and will be selected.
+  ///
+  /// \param DevSelector is a reference to a device_selector instance.
+  explicit PiMock(const cl::sycl::device_selector &DevSelector)
+      : PiMock(cl::sycl::platform{DevSelector}) {}
+
+  /// Explicit construction from a host_selector is forbidden.
+  PiMock(const cl::sycl::host_selector &HostSelector) = delete;
```

Comment:
And what happens if the default selector pick the host device?

Reply:
It looks the unit test handle this case anyway. So it is not technically forbidden?

Reply:
In this case, the assertion would fire. The unit test handles the case in order to duck the assertion in a host-only environment.
Would it be better to run `default_selector::select_device()` within the constructor?

Reply:
No need for it since it will assert on line 113 already.
Here I understand it is only to trigger a `static_assert` instead of an assert.

Reply:
Ah, yes, that's the point - disallowing statically whenever possible. Got confused myself with my previous suggestion :D